Created `TodoController` to handle the business logic of the application

AJThePro99 opened the pull request:
Added attributes:
- `id`
- `lastId`

Added declarations of:
- `checkTodo()`
- `uncheckTodo()`
- `listTodo()`
- `deleteTodo()`
- `peekTodo()`

- `editTodo()` -> `TodoModel` return type
- `createTodo()` -> `TodoModel` 

---

Have to implement:
- Function definitions
- Unique id system, and functions around it 
- Integration with `main.cpp`

AJThePro99 committed:
Creaded class files for further development of the TodoControl feature.

AJThePro99 committed:
Creaded class files for further development of the TodoControl feature.

AJThePro99 commented:
@lokesh-tr Please review and comment whether the progress is headed in the right direction. And, do tell me if I'm missing anything in there.

lokesh-tr commented on include/todo_lib/TodoControl.h:
```diff
@@ -0,0 +1,46 @@
+
+#include "todo_lib/TodoModel.h"
+
+class TodoControl {
```
I would be happy if you rename the file and class and header to `TodoController`. Its a convention

lokesh-tr commented on include/todo_lib/TodoControl.h:
```diff
@@ -0,0 +1,46 @@
+
+class TodoControl {
+private:
+    unsigned int id;
```
I dunno why you need two types of id here. I guess one would be enough.

As far as the properties are concerned, you will also need an array to store the `TodoModel`s.
And try to implement this as a singleton, for which you might need a property called `instance`.

I would recommend you to look into how to implement singletons in C++.

lokesh-tr commented on include/todo_lib/TodoControl.h:
```diff
@@ -0,0 +1,46 @@
+public:    
+    void checkTodo();
+    void uncheckTodo();
+    void listTodo();
```
`listTodo` must return an array of `TodoModel`.

lokesh-tr replied:
By array, I mean it could be a `std::vector`

lokesh-tr commented on include/todo_lib/TodoControl.h:
```diff
@@ -0,0 +1,46 @@
+    static int lastId;
+
+public:    
+    void checkTodo();
```
`checkTodo`, `uncheckTodo`, `peekTodo`, `deleteTodo`, `editTodo` should all take an `id` as argument.

lokesh-tr commented on include/todo_lib/TodoControl.h:
```diff
@@ -0,0 +1,46 @@
+    void uncheckTodo();
+    void listTodo();
+    void deleteTodo();
+    void peekTodo();
```
`peekTodo` should take in an `id` and return a `TodoModel`

lokesh-tr commented on include/todo_lib/TodoControl.h:
```diff
@@ -0,0 +1,46 @@
+    void deleteTodo();
+    void peekTodo();
+
+    TodoModel editTodo();
```
`editTodo` and `createTodo` need not return a `TodoModel`, they return an int representing the `id`

lokesh-tr commented on src/TodoControl.cpp:
```diff
@@ -0,0 +1,41 @@
+
+#include <todo_lib\TodoControl.h>
+
+void TodoControl::checkTodo() {
```
Change these according to the new function signatures which i suggested.

lokesh-tr reviewed (commented):
I think we will be heading in the right direction if you make the appropriate changes.

lokesh-tr commented:
And you need not worry about integration with main.cpp in this PR

lokesh-tr commented:
The `std::vector<TodoModel>` attribute is the only thing which i find missing apart from the other changes which i have mentioned.

lokesh-tr committed:
Implemented storing and retrieving `int lastId` and `std::vector<TodoModel>` into `data.json` file.

Changes `CMakeLists.txt` to import nlohmann/json library for JSON Serialization & Deserialization.

lokesh-tr committed:
Merge pull request #9 from lokesh-tr/lokesh-tr/json-and-persistence

Implemented storing and retrieving `int lastId` and `std::vector<TodoModel>` into `data.json` file.

AJThePro99 committed:
Added implementation of functions inf TodoController

AJThePro99 committed:
Added implementation of functions inf TodoController

lokesh-tr committed:
Create CODE_OF_CONDUCT.md

AJThePro99 commented:
@lokesh-tr  I have implemented all the necessary functions to the best of my knowledge. I still am not sure about the singleton function implementation, since I do not have much knowledge of it yet. But I've managed to do everything else so far.

lokesh-tr commented on include/todo_lib/TodoController.h:
```diff
@@ -0,0 +1,43 @@
+//
+// ===-----------------------------------------------------------------------------------------===
+
+#ifndef TODOCONTROL_H
```
make this `TODOCONTROLLER_H`

lokesh-tr commented on include/todo_lib/TodoController.h:
```diff
@@ -0,0 +1,43 @@
+    std::vector<TodoModel> listTodo();
+
+    int editTodo(int);
+    int createTodo();
```
Add argument names to function signatures.
for example:
`void checkTodo(int id)`

lokesh-tr replied:
`createTodo` should take `title` as required arguments, `description`, and `status` as optional arguments

lokesh-tr commented on include/todo_lib/TodoController.h:
```diff
@@ -0,0 +1,43 @@
+    TodoModel peekTodo(int);
+    std::vector<TodoModel> listTodo();
+
+    int editTodo(int);
```
`editTodo` should take `id` as a required argument and `title`, `description`, `status` as optional arguments

AJThePro99 replied:
If all the three attributes (`title`, `description`, `status`) are optional, what function will the `editTodo(int id)` will be performing? And, would it return an int? (returning `id`)

lokesh-tr commented on src/TodoController.cpp:
```diff
@@ -0,0 +1,82 @@
+            return todo;
+        }
+    }
+    throw std::invalid_argument("Todo not found");
```
Don't throw error, return null

lokesh-tr commented on src/TodoController.cpp:
```diff
@@ -0,0 +1,82 @@
+}
+
+void TodoController::deleteTodo(int id) {
+    todos.erase(std::remove_if(todos.begin(), todos.end(), [id](TodoModel& todo) {
+        return todo.getId() == id;
+    }), todos.end());
```
Explain me this piece of code

lokesh-tr commented on src/TodoController.cpp:
```diff
@@ -0,0 +1,82 @@
+    throw std::invalid_argument("Todo not found");
+}
+
+int TodoController::editTodo(int id) {
+    for (auto &todo : todos) {
+        if (todo.getId() == id) {
+            std::string filler;
+            std::cout << "Title: ";
+            std::cin.ignore();
+            std::getline(std::cin, filler);
+            todo.setTitle(filler);
+
+            std::cout << "Description: ";
+            std::cin.ignore();
+            std::getline(std::cin, filler);
+            todo.setDescription(filler);
+            return id;
+        }
+    }
+    throw std::invalid_argument("Todo not found");
+}
```
Get the values as parameters not as input from the console

lokesh-tr commented on src/TodoController.cpp:
```diff
@@ -0,0 +1,82 @@
+    throw std::invalid_argument("Todo not found");
+}
+
+int TodoController::createTodo() {
+    TodoModel todo = TodoModel(++lastId);
+    todos.push_back(todo);
+    return lastId;
+}
```
Same thing here, get values as parameters and pass it to constructor along with the id.

This might also mean that your constructor definition in `TodoModel` should take `id`, `title` as required argument and `description`, `status` as optional arguments

lokesh-tr commented on src/TodoModel.cpp:
```diff
@@ -20,6 +20,12 @@ TodoModel::TodoModel() {
     this->description = "Description ...";
     this->status = false;
 }
+TodoModel::TodoModel(int id) {
```
Yeah so, in a nutshell,
A TodoModel can only be able to be constructed if and only if an `id` and `title` is supplied. but, we should be happy to receive `description` and `status` but its not mandatory. The default values for `description` and `status` if not supplied should be `""` and `false` respectively.

Make appropriate changes in the header file too.

lokesh-tr commented on include/todo_lib/TodoModel.h:
```diff
@@ -23,6 +23,7 @@ class TodoModel {
     std::string description;
 public:
     TodoModel();
+    TodoModel(int id);
     TodoModel(std::string title);
     TodoModel(std::string title, std::string description);
     TodoModel(std::string title, std::string description, bool status);
```
`id` and `title` should be mandatory.

lokesh-tr commented on include/todo_lib/TodoModel.h:
```diff
@@ -23,6 +23,7 @@ class TodoModel {
     std::string description;
 public:
     TodoModel();
```
remove the no argument constructor.

lokesh-tr reviewed (changes requested):
Thanks @AJThePro99, I have a few changes otherwise it looks good to me. I advise you to read the full review and then start working on the fixes one at a time.

The thing which i want you to be careful about is that there should be no way you can create a `TodoModel` if an `id` and `title` are not supplied.

And in the `TodoController`, don't try to interface with the console I/O, get inputs as parameters.

AJThePro99 committed:
Finished implementatin of TodoController class, along with fixes in main.cpp and CMakeLists.txt to allow compilation

AJThePro99 committed:
Finished implementatin of TodoController class, along with fixes in main.cpp and CMakeLists.txt to allow compilation

AJThePro99 committed:
Merge branch 'todo-control-class' of https://github.com/AJThePro99/todo into todo-control-class

AJThePro99 committed:
Fixed backward slash issue - should work on linux and macOS for now
